### PR TITLE
[master] Bug 571794: Enable Input parameters for Criteria API SELECT clause

### PR DESCRIPTION
**Pull request by dazey3**
for #1034

Signed-off-by: Will Dazey <dazeydev.3@gmail.com>

### CODE REVIEW

**Comment by dazey3** [comment]
I have this PR as a draft so that it doesn't get merged without the reviewer reading https://bugs.eclipse.org/bugs/show_bug.cgi?id=571794 first. There may be additional changes that should be brought up in the JPA Spec git if this change is accepted.

**Review by lukasj — COMMENTED** [review]
@dazey3 The most important question - do you need this, if approved, in 3.0.1 or 2.7.x/2.6.x are more important and master can wait?

In any case, let me paste here your statements from the bug and inline my comments for better readability:

> I'm not sure if this is a bug or a feature, but the code and JPQL behavior implies to me it's a bug even though the JPA Spec seems to indicate this shouldn't work:

As I understand it the spec does not indicate it should nor it should not work. As you pointed out, following is there in the spec:

```
4.6.4 Input Parameters

    Input parameters can only be used in the WHERE clause or HAVING clause of a query or as the new value for an update item in the SET clause of an update statement.
```

Per my understanding, all it says is that in given context, input parameters MUST be supported. It says nothing about what SHOULD or SHOULD NOT happen in other cases, ie the one you're dealing with, so based on that, I believe that the behaviour of input parameter in the SELECT clause is undefined by the spec and thus - from that perspective - current behaviour in both cases is compliant with the spec itself. (There are few more restrictions on input parameters through the spec but I believe none of them is applicable to this particular context)

> TLDR, I have two examples: a JPQL and a Criteria API example. Both are equivalent, but the JPQL runs just fine and the Criteria API fails.

...one is likely a bug, the other a feature. But which one is right and which one is wrong? :-)


> The JPA spec seems to say that neither are valid and it should error!

I do not agree with you here, I think the spec is silent on what exactly should happen in this particular case (as mentioned above)

> My opinion is that since it works in JPQL, keeping Criteria API in sync is more important.

Agree!

> If we want to adhere to the JPA Spec, then we should alter JPQL to throw an exception and I am not in favor of removing this functionality given that databases support it. If anything, the JPA spec should expand to include input parameters in SELECT clause. Thoughts?
 
I would probably not change the spec here and keep it open for interpretation unless there is broader agreement across other vendors that this needs to be addressed.

**Comment by lukasj** [comment]
btw: to make it clear - I'm for approving the change `as is`

**Comment by dazey3** [comment]
@lukasj 
Thanks for looking at this!

> The most important question - do you need this, if approved, in 3.0.1 or 2.7.x/2.6.x are more important and master can wait?

It can wait. 

> As I understand it the spec does not indicate it should, nor it should not work.
> Per my understanding, all it says is that in given context, input parameters MUST be supported.

```
4.6.4 Input Parameters

    Input parameters can only be used in the WHERE clause or HAVING clause of a query or as the new value for an update item in the SET clause of an update statement.
```

I disagree with that interpretation of 4.6.4. It isn't just saying that input parameters are supported in WHERE / HAVING clauses, to me, it's saying the JPA spec ONLY supports the use of input parameters in WHERE / HAVING clauses. That's a pretty big distinction and was the root of my dilemma with just "fixing" this without saying anything. If you "can only do X", then I think that prohibits you from doing Y if Y does not exist as part of X. 

> I would probably not change the spec here and keep it open for interpretation unless there is broader agreement across other vendors that this needs to be addressed.

I'm generally not a fan of keeping things vague when it can be clarified. I have seen customers desired use of parameters in SELECT statements (stand alone, in a constructor, or a case expression) and it seems like this is simple enough to expand the JPA Spec to say "Only in SELECT, WHERE, or HAVING clauses"

**Comment by lukasj** [comment]
>  it's saying the JPA spec ONLY supports the use of input parameters in WHERE / HAVING clauses.

Right. But the spec does not explicitly say, what should happen if input parameters are used on other places. I think it should be read as "what is not forbidden, is allowed" - providers are free to either support such usage or fail and they should be consistent in the behaviour they choose (which was not the case before this PR)

> I'm generally not a fan of keeping things vague when it can be clarified.

the best thing to do is to file an issue in the jpa-api project and deal with it there. I am by no means putting myself into the role of the only decision maker ;-)

**Comment by dazey3** [comment]
Ok, I guess we both read it differently. To me it hinges on the word "only" and to me, the use of that word means it prohibits the use outside of WHERE and HAVING. I agree that the spec does not expressly define what happens in the case where an input parameter is using outside of WHERE or HAVING clauses, but I would think that is because outside of WHERE and HAVING is invalid/prohibited. The spec doesnt describe specific error messages much.

I also created a PR for 3.0 and 2.7. If you want to approve them, I am good with merging. I think we both agree the change is good, even if you feel it is EclipseLink specific behavior. I just wanted to make sure it was not in direct opposition to the spec.

**Review by lukasj — APPROVED** [review]
LGTM
